source/common/router/router.cc

```diff
@@ -1052,7 +1052,7 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& head
   // chance to return before returning a response downstream.
   if (could_not_retry && (numRequestsAwaitingHeaders() > 0 || pending_retries_ > 0)) {
     upstream_request.upstream_host_->stats().rq_error_.inc();
-    upstream_request.removeFromList(upstream_requests_);
+    upstream_request.removeFromList(upstream_requests_)->resetStream();
```

Comment:
Sorry QQ in the non-hedging case, how did this work previously? Where were things reset, and how do we avoid resetting things twice now?

Reply:
In the non-hedging case you'd fall through to `Filter::onDestroy()` which resets all remaining upstream requests.

It won't be reset twice because we're removing it from the upstream request list on this line, so it won't be seen in `resetAll()`.

Reply:
OK can you maybe just add as mall comment about this? Makes sense.

/wait